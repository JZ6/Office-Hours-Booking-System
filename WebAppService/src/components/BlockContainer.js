import React from "react";
import moment from "moment";

import "../styles/BlockContainer.css";
import BlockView from "./BlockView";
import SlotView from "./SlotView";

export default class BlockContainer extends React.Component {
	constructor(props) {
		super(props);
		
		this.state = {
			visible: false,
			locked: false,
			prevSlots: []
		}
		
		// Blocks
		this.handleInputChange = this.handleInputChange.bind(this);
		this.submitBlock = this.submitBlock.bind(this);
		this.updateBlock = this.updateBlock.bind(this);
		this.deleteBlock = this.deleteBlock.bind(this);
		this.onClose = this.onClose.bind(this)
		
		// Slots
		this.handleUpdate = this.handleUpdate.bind(this);
	}
	
	onOpen(block) {
		this.setState({
			visible: true,
			...block,
			start: moment(block.startTime).format("HH:mm"),
			end: this.getEnd(block),
			prevSlots: this.copySlots(block.appointmentSlots)
		});
	}
	onClose() {
		if (!this.state.locked && this.state.visible) {
			this.props.blockContainerClose();
			this.setState({
				visible: false,
			});
		}
	}
	
	update(scope, i) {
		if (!this.state.locked && this.state.visible) {
			this.setState({locked: true});
			this.props.api.getBlock(this.state.blockId)
			.then((response) => {
				if (response.status === 200) {
					// Successful, return json promise to next .then
					return response.json();
				} else {
					window.alert(response.status, response.statusText);
				}
				this.setState({locked: false});
			})
			.then((data) => {
				if (data.appointmentSlots) {
					// Extract data from json promise, undefined if failure
					if (scope === "block") {
						let prevSlots = this.copySlots(data.appointmentSlots);
						this.setState({prevSlots: prevSlots});
						this.setState({...data});
						
						this.setState({
							start: moment(data.startTime).format("HH:mm"),
							end: this.getEnd(data)
						});
					} else if (scope === "slots") {
						let prevSlots = this.copySlots(data.appointmentSlots);
						this.setState({prevSlots: prevSlots});
						this.setState({appointmentSlots: data.appointmentSlots});
						
						this.setState({end: this.getEnd(data)});
					} else if (scope === "slot") {
						console.log(this.state.prevSlots);
						
						let prevSlots = this.copySlots(this.state.prevSlots);
						prevSlots[i].identity = data.appointmentSlots[i].identity;
						prevSlots[i].note = data.appointmentSlots[i].note;
						this.setState({prevSlots: prevSlots});
						this.editSlot(
							i, 
							data.appointmentSlots[i].identity, 
							data.appointmentSlots[i].note
						);
						
						this.setState({end: this.getEnd(data)});
					}
				}
				this.setState({locked: false});
			})
			.catch((error) => {
				window.alert(error.message);
				this.setState({locked: false});
			});
		}
	}
	
	//----------------------------------------------------------------------------
	//---------------------------------- Block -----------------------------------
	//----------------------------------------------------------------------------
	
	getEnd(block) {
		return moment(block.startTime).add(block.appointmentDuration * block.appointmentSlots.length).format("HH:mm");
	}
	
	updateSlotNumber(start, end, appointmentDuration){
		if (!this.state.locked && this.state.visible) {
			let slotNumber = Math.floor(
				(moment(end, "HH:mm") - moment(start, "HH:mm")) / appointmentDuration);
			if (slotNumber === 0) {
				this.setState({prevSlots: []});
				this.setState({appointmentSlots: []});
			} else {
				let slots = [...Array(slotNumber)].map(() => 
					({identity: "", courseCode: "", note: ""}));
				this.setState({prevSlots: this.copySlots(slots)});
				this.setState({appointmentSlots: slots});
				
				// Update End Time?
				this.setState({
					end: moment(start, "HH:mm").add(appointmentDuration * slotNumber).format("HH:mm")
				});
			}
		}
	}
	
	handleInputChange(event) {
		if (!this.state.locked && this.state.visible) {
			const value = event.target.value;
			const name = event.target.name;
			
			if (name==="appointmentDuration") {
				let duration = parseInt(value, 10);
				if (duration && duration > 0) {
					this.updateSlotNumber(this.state.start, this.state.end, duration);
					this.setState({appointmentDuration: duration});
				}
			} else if (name === "start") {
				// Validate
				let start = value;
				if (moment(start, "HH:mm").isValid()) {
					if (moment(start, "HH:mm") > moment(this.state.end, "HH:mm")) {
						start = this.state.end;
					}
					this.setState({start: start});
					
					// Need to update block's actual startTime which includes date
					let newStartTime = this.state.startTime;  // String immutable
					newStartTime = moment(newStartTime)
						.hour(parseInt(start.slice(0, 2), 10))
						.minute(parseInt(start.slice(3, 5), 10))
						.second(0)
						.millisecond(0)
						.toISOString();
					this.setState({startTime: newStartTime})
					
					// Changes number of slots
					this.updateSlotNumber(start, this.state.end, this.state.appointmentDuration);
				}
			} else if (name === "end") { 
				// Validate
				let end = value;
				if (moment(end, "HH:mm").isValid()) {
					if (moment(end, "HH:mm") < moment(this.state.start, "HH:mm")) {
						end = this.state.start;
					}
					this.setState({end: end});
					
					// Changes number of slots
					this.updateSlotNumber(this.state.start, end, this.state.appointmentDuration);
				}
			} else if (name === "date") {
				// ValiDATE
				let date = value;
				if (moment(date, "YYYY-MM-DD").isValid()) {
					let newStartTime = moment(date, "YYYY-MM-DD")
						.add(moment(this.state.startTime).hour(), "hours")
						.add(moment(this.state.startTime).minute(), "minutes")
						.add(moment(this.state.startTime).second(), "seconds")
						.toISOString();
					this.setState({startTime:newStartTime})

					// Shouldn't change number of slots
				}
			} else if (name === "courseCodes") {
				this.setState({courseCodes: value.split(",")});
			} else if (name === "owners") {
				this.setState({owners: value.split(",")});
			} else if (name === "comment") {
				this.setState({comment: value});
			} else {
				window.alert("Invalid form element name ", name);
			}
		}
	}

	//update relevant selectedCourses when a course is selected
	handleCourseSelection(event){
		if (!this.state.locked && this.state.visible) {
			var newSelectedCourses = this.state.selectedCourses;
			var course = event.target.id;

			var target = event.target;
			const value = target.type === 'checkbox' ? target.checked : target.value;
			//add course to course list if checked
			if (value === true){
				if(!newSelectedCourses.includes(course)){
					//add course to selected courses
					newSelectedCourses.push(course);
					this.setState({selectedCourses:newSelectedCourses});
				}
			} else { //remove course from selected courses
				var index = newSelectedCourses.indexOf(course);
				if (index > -1) {
					newSelectedCourses.splice(index, 1);
				}
			}
			console.log(this.state.selectedCourses);
		}
	}
	
	submitBlock() {
		if (!this.state.locked && this.state.visible) {
			this.setState({locked: true});
			let block = {
				blockId: this.state.blockId,
				owners: this.state.owners,
				courseCodes: this.state.courseCodes,
				comment: this.state.comment,
				startTime: this.state.startTime,
				appointmentDuration: this.state.appointmentDuration,
				appointmentSlots: this.state.appointmentSlots
			}
			this.props.api.postBlock(block)
			.then((response) => {
				if (response.status !== 200) {
					window.alert(response.status, response.statusText);
				}
				this.setState({locked: false});
				this.update("block");
			})
			.catch((error) => {
				window.alert(error.message);
				this.setState({locked: false});
			});
			
			this.props.blockContainerCallback(this.state.blockId, block);
			this.onClose();
		}
	}
	
	updateBlock() {
		this.update("block");
	}
	
	deleteBlock() {
		if (!this.state.locked && this.state.visible) {
			if (!this.state.blockId) {
				this.onClose();
				return false;
			}
			this.setState({locked: true});
			this.props.api.deleteBlock(this.state.blockId)
			.then((response) => {
				if (response.status !== 200) {
					window.alert(response.status, response.statusText);
				}
				this.setState({locked: false});
			})
			.catch((error) => {
				window.alert(error.message);
				this.setState({locked: false});
			});
			
			this.props.blockContainerCallback(this.state.blockId);
			this.onClose();
		}
	}
	
	//----------------------------------------------------------------------------
	//---------------------------------- Slots -----------------------------------
	//----------------------------------------------------------------------------
	
	handleSlotClick = (i) => () => {
		if (!this.state.locked && this.state.visible) {
			if (this.state.appointmentSlots[i].identity === "") {
				// Click on empty slot to assign
				this.editSlot(i, this.props.id, "");
			} else {
				if (this.props.id === this.state.appointmentSlots[i].identity) {
					// Click on own slot to delete
					this.editSlot(i, "", "");
				}
				//Click on occupied slot to do nothing
			}
		}
	}
	
	handleIdentityChange = (i) => (event) => {
		if (!this.state.locked && this.state.visible) {
			// Delete note to protect privacy
			this.editSlot(i, event.target.value, "");
		}
	}
	
	handleNoteChange = (i) => (event) => {
		if (!this.state.locked && this.state.visible) {
			// Can't edit note that belongs to nobody
			if (this.state.appointmentSlots[i].identity !== "") {
				this.editSlot(i, this.state.appointmentSlots[i].identity, event.target.value);
			} else {
				window.alert("Can't edit note of an unregistered slot.");
			}
		}
	}
	
	handleSlotConfirm = (i) => () => {
		if (!this.state.locked && this.state.visible) {
			this.setState({locked: true});
			this.props.api.editSlot(
				this.state.blockId, i, 
				{
					identity: this.state.appointmentSlots[i].identity, 
					courseCode: this.state.appointmentSlots[i].courseCode, 
					note: this.state.appointmentSlots[i].note
				})
			.then((response) => {
				if (response.status !== 200) {
					window.alert(response.status, response.statusText);
				}
				this.setState({locked: false});
				this.update("slot", i);
			})
			.catch((error) => {
				window.alert(error.message);
				this.setState({locked: false});
				this.update("slot", i);
			});
		}
	}
	
	handleSlotCancel = (i) => () => {
		this.update("slot", i);
	}
	
	handleEmpty = () => {
		if (!this.state.locked && this.state.visible) {
			this.setState({locked: true});
			let promises = [];
			
			if (this.props.role === "student") {
				// Student clears only their own
				for (let i = 0; i < this.state.appointmentSlots.length; i++) {
					if (this.props.id === this.state.appointmentSlots[i].identity) {
						promises.push(this.props.api.editSlot(
							this.state.blockId, i, 
							{identity: "", courseCode: "", note: ""}));
					}
				}
			} else {
				// Instructor clears all
				for (let i = 0; i < this.state.appointmentSlots.length; i++) {
					promises.push(this.props.api.editSlot(
						this.state.blockId, i, 
						{identity: "", courseCode: "", note: ""}));
				}
			}
			Promise.all(promises)
			.then((responses) => {
				responses.forEach((response) => {
					if (response.status !== 200) {
						window.alert(response.status, response.statusText);
					}
				});
				this.setState({locked: false});
				this.update("slots");
			})
			.catch((error) => {
				window.alert(error.message);
				this.setState({locked: false});
				this.update("slots");
			})
		}
	}
	
	handleUpdate() {
		this.update("slots");
	}
	
	editSlot(i, identity, note) {
		const newSlots = this.copySlots(this.state.appointmentSlots);
		newSlots[i].identity = identity;
		newSlots[i].note = note;
		this.setState({appointmentSlots: newSlots});
	}
	
	copySlots(slots) {
		return slots.map(slot => {
			return {
				identity: slot.identity, 
				courseCode: slot.courseCode, 
				note: slot.note
			}
		});
	}
	
	//----------------------------------------------------------------------------
	//---------------------------------- Render ----------------------------------
	//----------------------------------------------------------------------------
	
	render() {
		if (this.state.visible) {
			return <div className="BlockContainer">Loading: {(this.state.locked).toString()}
				<BlockView 
					handleInputChange={this.handleInputChange}
					submitBlock={this.submitBlock}
					updateBlock={this.updateBlock}
					deleteBlock={this.deleteBlock}
					onClose={this.onClose}
					
					startTime={this.state.startTime}
					appointmentDuration={this.state.appointmentDuration}
					owners={this.state.owners}
					courseCodes={this.state.courseCodes}
					comment={this.state.comment}
					appointmentSlots={this.state.appointmentSlots}
					blockId={this.state.blockId}
					
					start={this.state.start}
					end={this.state.end}
					
					role={this.props.role}
					id={this.props.id}
				/>
				{!this.state.blockId ? 
					"Please submit the new block before editing slots."
				: 
					<SlotView 
						handleSlotClick={this.handleSlotClick}
						handleIdentityChange={this.handleIdentityChange}
						handleNoteChange={this.handleNoteChange}
						handleSlotConfirm={this.handleSlotConfirm}
						handleSlotCancel={this.handleSlotCancel}
						handleEmpty={this.handleEmpty}
						handleUpdate={this.handleUpdate}
						
						startTime={this.state.startTime}
						slotDuration={this.state.appointmentDuration}
						slots={this.state.appointmentSlots}
						prevSlots={this.state.prevSlots}
						role={this.props.role}
						id={this.props.id}
					/>
				}
			</div>;
		}
		else {
			return null;
		}
	}
}
