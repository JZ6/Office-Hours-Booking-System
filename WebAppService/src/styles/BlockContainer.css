
.BlockContainer{
	position: absolute;
	width: 80%;
	height: 90%;
	top:50%;
	left:50%;
	transform: translate(-50%,-50%);
	background-color:rgb(255, 255, 255);
	z-index: 6;
	border-radius: 4px;
	box-shadow: 0 8px 8px 0 rgba(0, 0, 0, 0.4);
	padding-top: 1em;
	display: flex;
	flex-wrap: wrap;
	flex-flow: column;
}

.BlockContainer button {
	background-color: dodgerblue;
	color: white;
	font-size: 16px;
	border-radius: 4px;
	padding: 4px 8px;
	min-width: 150px;
	border: none;
	margin: 4px 4px 8px 4px;
	box-shadow: 0 4px rgb(0, 89, 179);
}
.BlockContainer button:hover {
	background-color: rgb(77, 166, 255);
	transform: translateY(1px);
	box-shadow: 0 3px rgb(0, 89, 179);
}
.BlockContainer button:active {
	background-color: rgb(77, 166, 255);
	transform: translateY(4px);
	box-shadow: 0 0px rgb(0, 89, 179);
}
.BlockContainer input{
	border-radius: 4px;
	border: 1px solid lightgray;
	padding: 4px 8px;
	display: inline-block;
	margin: 4px;
}
.BlockContainer textarea{
	border-radius: 4px;
	border: 1px solid lightgray;
	padding: 4px 8px;
	margin: 4px;
	height: 50%;
	resize: none;
}

.BlockView{
	display: flex;
	width: 100%;
	height: 100px;
}
.blockInfo{
	width: 50%;
	height: 100%;
}
.blockTimes{
	width: 50%;
	height: 100%;
}
.blockTimes>div{
}

.ButtonContainer{
	width: 100%;
	height: 38px;
}
