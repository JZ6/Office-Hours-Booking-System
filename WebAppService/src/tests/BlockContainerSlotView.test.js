import React from "react";
import renderer from 'react-test-renderer';
import {shallow, mount} from "enzyme";
import moment from "moment";
import BlockContainer from "../components/BlockContainer";

let mockJson, mockGetBlock, mockEditSlot;
let mockJsonPromise, mockGetBlockPromise, mockEditSlotPromise;
let mockProps, mockBlock, mockApi;

beforeEach(() => {
	mockBlock = {
		blockId: "0",
		owners: ["rossbob2", "atat4"],
		courseCodes: ["csc301", "csc302"],
		comment: "We don't make mistakes, just happy little accidents.",
		startTime: "2018-11-18T12:00:00Z",
		appointmentDuration: 600000,
		appointmentSlots: [
			{identity: "parkerpeter15", note: "Everyone gets one.", courseCode: "csc302"},
			{identity: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!", courseCode: "csc302"},
			{identity: "", note: "", courseCode: ""}
		]
	};
	mockJsonPromise = Promise.resolve(mockBlock);
	mockJson = jest.fn(() => mockJsonPromise);
	mockGetBlockPromise = Promise.resolve({
		status: 200,
		statusText: "OK",
		json: mockJson
	});
	mockGetBlock = jest.fn((blockId) => mockGetBlockPromise);
	mockEditSlotPromise = Promise.resolve({
		status: 200,
		statusText: "OK"
	});
	mockEditSlot = jest.fn((blockId, slotId, slot) => {
		mockBlock.appointmentSlots[slotId] = slot;
		return mockEditSlotPromise;
	});
	mockApi = {
		getBlock: mockGetBlock,
		editSlot: mockEditSlot
		};
	mockProps = {
		api: mockApi
	};
});

describe("instructor view", () => {
	let wrapper;
	beforeEach(() => {
		mockProps.id = "octaviusotto3";
		mockProps.role = "instructor";
		wrapper = mount(<BlockContainer {...mockProps}/>);
		wrapper.instance().onOpen(mockBlock);
		wrapper = wrapper.update();
	});
	
	test("render", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});
	
	test("handleSlotClick isn't called", () => {
		wrapper.find("#identity2").simulate("click");
		expect(wrapper.find("#identity2").props().value).toEqual("");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
	
	test("handleIdentityChange changes identity", () => {
		// Edit empty
		wrapper.find("#identity2").simulate("change", {target: {value: "osbornharry31"}});
		expect(wrapper.find("#identity2").props().value).toEqual("osbornharry31");
		// Edit occupied (should clear note)
		wrapper.find("#identity0").simulate("change", {target: {value: "osbornharry31"}});
		expect(wrapper.find("#identity0").props().value).toEqual("osbornharry31");
		expect(wrapper.find("#note0").props().value).toEqual("");
	});
	
	test("handleNoteChange changes note", () => {
		// Edit empty (shouldn't allow changes)
		wrapper.find("#note2").simulate("change", {target: {value: "Honor is for fools."}});
		expect(wrapper.find("#note2").props().value).toEqual("");
		// Edit occupied
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		expect(wrapper.find("#note0").props().value).toEqual("Honor is for fools.");
	});
	
	test("handleSlotConfirm applies slot changes", async () => {
		wrapper.find("#identity0").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#confirm0").simulate("click");
		
		await mockEditSlotPromise;
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockEditSlot).toHaveBeenCalledTimes(1);
		expect(mockEditSlot).toHaveBeenCalledWith("0", 0, {courseCode: "csc302", identity: "osbornharry31", note: "Honor is for fools."});
		expect(wrapper.find("#identity0").props().value).toEqual("osbornharry31");
		expect(wrapper.find("#note0").props().value).toEqual("Honor is for fools.");
	});
	
	test("handleSlotCancel cancels changes", () => {
		wrapper.find("#identity0").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#cancel0").simulate("click");
		
		expect(wrapper.find("#identity0").props().value).toEqual("parkerpeter15");
		expect(wrapper.find("#note0").props().value).toEqual("Everyone gets one.");
	});
	
	test("updateBlock updates slots", async () => {
		wrapper.find("#identity0").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#identity1").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note1").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#identity2").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note2").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#refresh-button").simulate("click");
		
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockGetBlock).toHaveBeenCalledTimes(1);
		expect(mockGetBlock).toHaveBeenCalledWith("0");
		expect(wrapper.find("#identity0").props().value).toEqual("parkerpeter15");
		expect(wrapper.find("#note0").props().value).toEqual("Everyone gets one.");
		expect(wrapper.find("#identity1").props().value).toEqual("watsonmj25");
		expect(wrapper.find("#note1").props().value).toEqual("Face it Tiger... you just hit the jackpot!");
		expect(wrapper.find("#identity2").props().value).toEqual("");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
	
	test("handleEmpty empties slots", async () => {
		wrapper.find("#empty-button").simulate("click");
		
		await mockEditSlotPromise;
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockEditSlot).toHaveBeenCalledTimes(3);
		expect(mockEditSlot).toHaveBeenCalledWith("0", 2, 
			{identity: "", courseCode: "", note: ""});
		expect(wrapper.find("#identity0").props().value).toEqual("");
		expect(wrapper.find("#note0").props().value).toEqual("");
		expect(wrapper.find("#identity1").props().value).toEqual("");
		expect(wrapper.find("#note1").props().value).toEqual("");
		expect(wrapper.find("#identity2").props().value).toEqual("");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
});

describe("student view", () => {
	let wrapper;
	beforeEach(async () => {
		mockProps.id = "parkerpeter15";
		mockProps.role = "student";
		wrapper = mount(<BlockContainer {...mockProps}/>);
		wrapper.instance().onOpen(mockBlock);
		wrapper = wrapper.update();
	});
	
	test("render", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});
	
	test("handleSlotClick assigns slot", () => {
		// Click on own slot (unregisters slot and deletes note)
		wrapper.find("#identity0").simulate("click");
		expect(wrapper.find("#identity0").props().children).toEqual("Register");
		expect(wrapper.find("#note0").props().value).toEqual(undefined);
		// Click on occupied slot (does nothing)
		wrapper.find("#identity1").simulate("click");
		expect(wrapper.find("#identity1").props().children).toEqual("Not Available");
		expect(wrapper.find("#note1").props().value).toEqual(undefined);
		// Click on empty slot (registers slot)
		wrapper.find("#identity2").simulate("click");
		expect(wrapper.find("#identity2").props().children).toEqual("Unregister");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
	
	test("handleNoteChange changes only own notes", () => {
		// Edit empty (shouldn't allow changes)
		expect(wrapper.find("#note2").props().onChange).toEqual(undefined);
		// Edit occupied (shouldn't allow changes)
		expect(wrapper.find("#note1").props().onChange).toEqual(undefined);
		// Edit own
		wrapper.find("#note0").simulate("change", {target: {value: "With great powers comes great responsibility."}});
		expect(wrapper.find("#note0").props().value).toEqual("With great powers comes great responsibility.");
	});
	
	test("handleSlotConfirm applies slot changes", async () => {
		wrapper.find("#identity2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#confirm2").simulate("click");
		
		await mockEditSlotPromise;
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockEditSlot).toHaveBeenCalledTimes(1);
		expect(mockEditSlot).toHaveBeenCalledWith("0", 2, {identity: "parkerpeter15", courseCode: "", note: "With great powers..."});
		expect(wrapper.find("#identity2").props().children).toEqual("Unregister");
		expect(wrapper.find("#note2").props().value).toEqual("With great powers...");
	});
	
	test("handleSlotCancel cancels changes", () => {
		// Edit identity and note
		wrapper.find("#identity2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#cancel2").simulate("click");
		
		expect(wrapper.find("#identity2").props().children).toEqual("Register");
		expect(wrapper.find("#note2").props().value).toEqual(undefined);
	});
	
	test("updateBlock updates slots", async () => {
		wrapper.find("#identity0").simulate("click");
		wrapper.find("#identity2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#refresh-button").simulate("click");
		
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockGetBlock).toHaveBeenCalledTimes(1);
		expect(mockGetBlock).toHaveBeenCalledWith("0");
		expect(wrapper.find("#identity0").props().children).toEqual("Unregister");
		expect(wrapper.find("#note0").props().value).toEqual("Everyone gets one.");
		expect(wrapper.find("#identity2").props().children).toEqual("Register");
		expect(wrapper.find("#note2").props().value).toEqual(undefined);
	});
	
	test("handleEmpty empties only own slots", async () => {
		wrapper.find("#empty-button").simulate("click");
		
		await mockEditSlotPromise;
		await mockGetBlockPromise;
		await mockJsonPromise;
		wrapper = wrapper.update();
		
		expect(mockEditSlot).toHaveBeenCalledTimes(1);
		expect(mockEditSlot).toHaveBeenCalledWith("0", 0, 
			{identity: "", courseCode: "", note: ""});
		expect(wrapper.find("#identity0").props().children).toEqual("Register");
		expect(wrapper.find("#note0").props().value).toEqual(undefined);
		expect(wrapper.find("#identity1").props().children).toEqual("Not Available");
		expect(wrapper.find("#note1").props().value).toEqual(undefined);
	});
});
