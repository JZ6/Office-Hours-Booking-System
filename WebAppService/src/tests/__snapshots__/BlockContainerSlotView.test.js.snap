// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instructor view render 1`] = `"<div class=\\"BlockContainer\\">Loading: false<div class=\\"BlockView\\"><div class=\\"close-block\\">Close</div><div class=\\"BlockTimes\\"><div>Warning: Changing times will clear slots!</div><input name=\\"start\\" type=\\"time\\" value=\\"07:00\\">-<input name=\\"end\\" type=\\"time\\" value=\\"07:30\\"><input name=\\"date\\" type=\\"date\\" value=\\"2018-11-18\\"><label>Slot duration:<select name=\\"appointmentDuration\\"><option value=\\"300000\\">5 minutes</option><option value=\\"600000\\">10 minutes</option><option value=\\"900000\\">15 minutes</option></select></label>Slot number: 3</div><div class=\\"blockInfo\\">Owners:<input name=\\"owners\\" type=\\"text\\" value=\\"rossbob2,atat4\\"><br>Courses:<input name=\\"courseCodes\\" type=\\"text\\" value=\\"csc301,csc302\\"><br>Block Description<textarea name=\\"comment\\" rows=\\"4\\" cols=\\"50\\" placeholder=\\"Enter block description...\\">We don't make mistakes, just happy little accidents.</textarea><br><button class=\\"submit-button\\">Submit</button><button class=\\"submit-button\\">Delete</button></div></div><div class=\\"slot-container\\"><button id=\\"empty-button\\">Empty All Slots</button><button id=\\"update-button\\">Refresh</button><div class=\\"Slot\\" id=\\"slot0\\">7:00AM - 7:10AM<input class=\\"text-input\\" name=\\"identity0\\" id=\\"identity0\\" type=\\"text\\" placeholder=\\"Unassigned UtorID...\\" maxlength=\\"50\\" value=\\"parkerpeter15\\"><input class=\\"text-input\\" name=\\"note0\\" id=\\"note0\\" type=\\"text\\" placeholder=\\"Empty Note...\\" maxlength=\\"280\\" value=\\"Everyone gets one.\\"><span></span></div><div class=\\"Slot\\" id=\\"slot1\\">7:10AM - 7:20AM<input class=\\"text-input\\" name=\\"identity1\\" id=\\"identity1\\" type=\\"text\\" placeholder=\\"Unassigned UtorID...\\" maxlength=\\"50\\" value=\\"watsonmj25\\"><input class=\\"text-input\\" name=\\"note1\\" id=\\"note1\\" type=\\"text\\" placeholder=\\"Empty Note...\\" maxlength=\\"280\\" value=\\"Face it Tiger... you just hit the jackpot!\\"><span></span></div><div class=\\"Slot\\" id=\\"slot2\\">7:20AM - 7:30AM<input class=\\"text-input\\" name=\\"identity2\\" id=\\"identity2\\" type=\\"text\\" placeholder=\\"Unassigned UtorID...\\" maxlength=\\"50\\" value=\\"\\"><input class=\\"text-input\\" name=\\"note2\\" id=\\"note2\\" type=\\"text\\" placeholder=\\"Empty Note...\\" maxlength=\\"280\\" value=\\"\\"><span></span></div></div></div>"`;

exports[`student view render 1`] = `"<div class=\\"BlockContainer\\">Loading: false<div class=\\"BlockView\\"><div class=\\"close-block\\">Close</div><div class=\\"blockTimes\\">From: 07:00AM, To: 07:30AM, Date: 2018-11-18</div><div class=\\"blockInfo\\">Owners: rossbob2atat4<br>Courses:csc301csc302<br><h3>Block Description</h3><div> We don't make mistakes, just happy little accidents.  </div></div></div><div class=\\"slot-container\\"><button id=\\"empty-button\\">Empty My Slots</button><button id=\\"update-button\\">Refresh</button><div class=\\"Slot--mine\\" id=\\"slot0\\">7:00AM - 7:10AM<button id=\\"identity0\\">Unregister</button><input class=\\"text-input\\" name=\\"note0\\" id=\\"note0\\" type=\\"text\\" placeholder=\\"Empty Note...\\" maxlength=\\"280\\" value=\\"Everyone gets one.\\"><span></span></div><div class=\\"Slot--taken\\" id=\\"slot1\\">7:10AM - 7:20AM<button id=\\"identity1\\" disabled=\\"\\">Not Available</button><span id=\\"note1\\"></span><span></span></div><div class=\\"Slot\\" id=\\"slot2\\">7:20AM - 7:30AM<button id=\\"identity2\\">Register</button><span id=\\"note2\\"></span><span></span></div></div></div>"`;
